Makefile-v2.mk

```diff
@@ -0,0 +1,296 @@
+## Build, test, and generate code for various parts of Grafana Agent.
+##
+## At least Go 1.18, git, and a moderately recent version of Docker is required
+## to be able to use the Makefile. If you do not have the full list of build
+## dependencies, you may set USE_CONTAINER=1 to proxy build commands to a build
+## container.
+##
+## Other environment variables can be used to tweak behaviors of targets.
+## See the bottom of this help section for the full list of supported
+## environment variables.
+##
+## Usage:
+##   make <target>
+##
+## Targets for running tests:
+##
+##   test  Run tests
+##   lint  Lint code
+##
+## Targets for building binaries:
+##
+##   binaries  Compiles all binaries.
+##   agent     Compiles cmd/agent to $(AGENT_BINARY)
+##   agentctl  Compiles cmd/agentctl to $(AGENTCTL_BINARY)
+##   operator  Compiles cmd/agent-operator to $(OPERATOR_BINARY)
+##   crow      Compiles tools/crow to $(CROW_BINARY)
+##   smoke     Compiles tools/smoke to $(SMOKE_BINARY)
+##
+## Targets for building Docker images:
+##
+##   images          Builds all Docker images.
+##   agent-image     Builds agent Docker image.
+##   agentctl-image  Builds agentctl Docker image.
+##   operator-image  Builds operator Docker image.
+##   crow-image      Builds crow Docker image.
+##   smoke-image     Builds smoke test Docker image.
+##
+## Targets for packaging:
+##
+##   dist                   Produce release assets for everything.
+##   dist-agent-binaries    Produce release-ready agent binaries.
+##   dist-agentctl-binaries Produce release-ready agentctl binaries.
+##   dist-packages          Produce release-ready DEB and RPM packages.
+##   dist-agent-intaller    Produce a Windows installer for Grafana Agent.
+##
+## Targets for generating assets:
+##
+##   generate             Generate everything.
+##   generate-crds        Generate Grafana Agent Operator CRDs.
+##   generate-manifests   Generate production/kubernetes YAML manifests.
+##   generate-dashboards  Generate dashboards in example/docker-compose after
+##                        changing Jsonnet.
+##   generate-protos      Generate protobuf files.
+##
+## Other targets:
+##
+##   build-container-cache  Create a cache for the build container to speed up
+##                          subsequent proxied builds
+##   clean                  Clean caches and built binaries
+##   help                   Displays this message
+##   info                   Print Makefile-specific environment variables
+##
+## Environment variables:
+##
+##   USE_CONTAINER    Set to 1 to enable proxying commands to build container
+##   AGENT_IMAGE      Image name:tag built by `make agent-image`
+##   AGENTCTL_IMAGE   Image name:tag built by `make agentctl-image`
```

Comment:
probably out of scope for this PR, but to simplify things a bit it seems we could combine agentctl, smoke, and crow images into a single agent tools image.

Reply:
Makes sense for the future (though I'm also tempted to combine agentctl + agent into a single binary by utilizing Cobra in Flow)